core/shark_turbine/aot/fx_programs.py

```diff
@@ -0,0 +1,325 @@
+    aliasing state and captured constants. Having those be physically shared
+    is an essential optimization.
+
+    In order to manage saving/loading of the set of things, we manually splice
+    the state_dict and constants dict such that while saving, we only persist
+    the first encountered instance of any reference. Any subsequent instances
+    are replaced with a SharedStateTensor, which on load can be re-associated.
+
+    As this is primarily targeted at being able to decouple FX tracing from
+    further manipulation (which for reasons unknown, is competing with the
+    race of entropy to the heat death of the universe in terms of performance),
+    we don't take a lot of pains to be optimized for distribution or storage of
+    the resulting artifacts.
+
+    In the future, this same technique could be employed to elide parameters
+    that we know we are going to resolve symbolically later, keeping them from
+    being loaded and consuming memory during model export and compilation.
+
+    We have faith that in the fullness of time, the design flaws in Torch that
+    require this kind of thing to exist will be resolved, and we then won't
+    need this hack.
+    """
+
+    def __init__(self):
+        self.programs: dict[str, torch.export.ExportedProgram] = {}
+
+    def save(self, path: Union[str, os.PathLike]) -> int:
+        """Saves the set of exported programs to a descriptor file.
+
+        Returns the number of tensors deduped (for debugging/testing).
+        """
+        path = Path(path).resolve()
+
+        def permute_path(name):
+            return path.parent / f"{path.stem}_{name}.pt2"
+
+        # Assemble descriptor.
+        program_files = {name: str(permute_path(name)) for name in self.programs.keys()}
+        descriptor = {
+            "load_order": list(program_files.keys()),
+            "program_files": program_files,
+        }
+
+        # Accumulate shared state as we go.
+        shared_state_dict: dict[str, Any] = {}
+        shared_constants: dict[str, Any] = {}
+        count_deduped = 0
+
+        # Save each.
+        for program_name, ep in self.programs.items():
+            # First validate the ep with normal rules, which we will then
+            # disable since we are violating the spec.
+            ep._validate()
+            orig_state_dict = dict(ep.state_dict)
+            constants_dict = _get_optional_constants(ep)
+            orig_constants = dict(constants_dict)
+
+            try:
+                # Now unmerge the state_dict and constants by knocking it up against
+                # our running shared state dict.
+                count_deduped += _sharify_state_dict(shared_state_dict, ep.state_dict)
+                count_deduped += _sharify_state_dict(shared_constants, constants_dict)
+
+                # And save our hacked program.
+                save_path = program_files[program_name]
+                torch.export.save(ep, save_path)
```

Comment:
So we are fundamentally saving a dict full of fakeTensors with metadata about what to load from our shared dict on load?

Reply:
nevermind, I see it in unsharify, we just check if the loaded value is a SharedStateTensor and redupe